test/lit/passes/optimize-instructions-gc.wast

```diff
@@ -533,4 +533,158 @@
       )
     )
   )
+
+  ;; CHECK:      (func $get-eqref (result eqref)
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $get-eqref (result eqref)
+    (unreachable)
+  )
+
+  ;; CHECK:      (func $ref-eq (param $x eqref) (param $y eqref)
+  ;; CHECK-NEXT:  (local $lx eqref)
+  ;; CHECK-NEXT:  (local $ly eqref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.eq
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:    (local.get $y)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.set $lx
+  ;; CHECK-NEXT:   (call $get-eqref)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $ref-eq (param $x eqref) (param $y eqref)
+    (local $lx eqref)
+    (local $ly eqref)
+    ;; identical parameters are equal
+    (drop
+      (ref.eq
+        (local.get $x)
+        (local.get $x)
+      )
+    )
+    ;; different ones might not be
+    (drop
+      (ref.eq
+        (local.get $x)
+        (local.get $y)
+      )
+    )
+    ;; identical locals are
+    (local.set $lx
+      (call $get-eqref)
+    )
+    (drop
+      (ref.eq
+        (local.get $lx)
+        (local.get $lx)
+      )
+    )
+    ;; fallthroughs work ok (but we need --remove-unused-names so that we can
+    ;; trivially tell that there are no breaks)
+    (drop
+      (ref.eq
+        (block (result eqref)
+          (nop)
+          (local.get $x)
+        )
+        (block (result eqref)
+          (nop)
+          (drop
+            (i32.const 10)
+          )
+          (nop)
+          (local.get $x)
+        )
+      )
+    )
+  )
+
+  (func $nothing)
+
+  ;; CHECK:      (func $ref-eq-corner-cases (param $x eqref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.eq
+  ;; CHECK-NEXT:    (block (result eqref)
+  ;; CHECK-NEXT:     (call $nothing)
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (block (result eqref)
+  ;; CHECK-NEXT:     (call $nothing)
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.eq
+  ;; CHECK-NEXT:    (struct.new_default_with_rtt $struct
+  ;; CHECK-NEXT:     (rtt.canon $struct)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (struct.new_default_with_rtt $struct
+  ;; CHECK-NEXT:     (rtt.canon $struct)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $ref-eq-corner-cases (param $x eqref)
+    ;; side effects prevent optimization
+    (drop
+      (ref.eq
+        (block (result eqref)
+          (call $nothing)
+          (local.get $x)
+        )
+        (block (result eqref)
+          (call $nothing)
+          (local.get $x)
+        )
+      )
+    )
+    ;; allocation prevents optimization
```

Comment:
It might be worth a test that dropped (or otherwise side-effect free) allocations do not prevent the optimization, since the fallthrough is done before the check for allocation instructions.

Reply:
Good idea, adding.